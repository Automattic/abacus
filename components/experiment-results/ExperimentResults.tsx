import _ from 'lodash'
import React, { useMemo } from 'react'

import { Analysis, ExperimentFull, MetricBare } from '@/models'

import CondensedLatestAnalyses from './CondensedLatestAnalyses'
import FullLatestAnalyses from './FullLatestAnalyses'
import ParticipantCounts from './ParticipantCounts'

/**
 * Main component for summarizing experiment results.
 */
export default function ExperimentResults({
  analyses,
  experiment,
  metrics,
  debugMode,
}: {
  analyses: Analysis[]
  experiment: ExperimentFull
  metrics: MetricBare[]
  debugMode?: boolean
}) {
  const metricsById = useMemo(() => _.zipObject(_.map(metrics, 'metricId'), metrics), [metrics])
  const metricAssignmentIdToLatestAnalyses = useMemo(
    () =>
      _.mapValues(_.groupBy(analyses, 'metricAssignmentId'), (metricAnalyses) => {
        metricAnalyses = _.orderBy(metricAnalyses, ['analysisDatetime'], ['desc'])
        return _.sortBy(
          _.filter(metricAnalyses, ['analysisDatetime', metricAnalyses[0].analysisDatetime]),
          'analysisStrategy',
        )
      }),
    [analyses],
  )

  if (analyses.length === 0) {
    return <p>No analyses yet for {experiment.name}.</p>
  }

  if (debugMode) {
    return (
      <>
        <p>Found {analyses.length} analysis objects in total.</p>

        <div className='analysis-participant-counts'>
          <h3>Participant counts for the primary metric</h3>
          <ParticipantCounts
            experiment={experiment}
            latestPrimaryMetricAnalyses={
              metricAssignmentIdToLatestAnalyses[experiment.getPrimaryMetricAssignmentId() as number]
            }
          />
        </div>

        <div className='analysis-latest-results'>
          <h3>Latest results by metric</h3>
          <FullLatestAnalyses
            experiment={experiment}
            metricsById={metricsById}
            metricAssignmentIdToLatestAnalyses={metricAssignmentIdToLatestAnalyses}
          />
        </div>

        <pre className='debug-json'>{JSON.stringify(analyses, null, 2)}</pre>
      </>
    )
  }

  return (
    <div className='analysis-latest-results'>
      <h3>Latest results by metric</h3>
      <CondensedLatestAnalyses
        experiment={experiment}
        metricsById={metricsById}
        metricAssignmentIdToLatestAnalyses={metricAssignmentIdToLatestAnalyses}
      />
    </div>
  )
}
