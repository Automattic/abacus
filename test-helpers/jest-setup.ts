import { format } from 'util'

// Throw errors on any console logging.
// Adapted from https://github.com/facebook/jest/issues/6121#issuecomment-529591574
;['error', 'info', 'log', 'warn'].forEach((funcName) => {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  ;(global.console as any)[funcName] = function (message?: any, ...optionalParams: any[]) {
    throw new Error(format(message, optionalParams))
  }
})
