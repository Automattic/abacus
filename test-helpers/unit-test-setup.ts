/* eslint-disable @typescript-eslint/ban-ts-ignore */
// @ts-ignore
window.URL.createObjectURL = jest.fn()

// @ts-ignore
window.HTMLCanvasElement.prototype.getContext = jest.fn()
