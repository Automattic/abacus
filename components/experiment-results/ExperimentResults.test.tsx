import { fireEvent, getAllByText, getByText, waitFor } from '@testing-library/react'
import React from 'react'

import ExperimentResults from '@/components/experiment-results/ExperimentResults'
import Fixtures from '@/test-helpers/fixtures'
import { render } from '@/test-helpers/test-utils'

const experiment = Fixtures.createExperimentFull()
const metrics = Fixtures.createMetricBares()
const analyses = Fixtures.createAnalyses()

test('renders an appropriate message with no analyses', () => {
  const { container } = render(<ExperimentResults analyses={[]} experiment={experiment} metrics={metrics} />)
  expect(container).toHaveTextContent('No analyses yet for experiment_1.')
})

test('renders the condensed table with some analyses in non-debug mode', async () => {
  const { container } = render(<ExperimentResults analyses={analyses} experiment={experiment} metrics={metrics} />)

  // In non-debug mode, we shouldn't have a <pre> element with the JSON.
  expect(container.querySelector('.debug-json')).toBeNull()

  // Check the table snapshot before expanding any metric.
  expect(container.querySelector('.analysis-latest-results')).toMatchSnapshot()

  // Clicking on metric_1 or metric_2 should have no effect on anything, but metric_3 should render the details.
  fireEvent.click(getByText(container, /metric_1/))
  fireEvent.click(getAllByText(container, /metric_2/)[0])
  fireEvent.click(getByText(container, /metric_3/))
  await waitFor(() => getByText(container, /Last analyzed/), { container })
  expect(container.querySelector('.analysis-latest-results .analysis-detail-panel')).toMatchSnapshot()
})
