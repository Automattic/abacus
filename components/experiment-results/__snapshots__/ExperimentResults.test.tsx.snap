// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the condensed table with some analyses in non-debug mode for a Conversion Metric 1`] = `
<div
  class="analysis-latest-results"
>
  <div
    class="makeStyles-root-1"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation2 MuiPaper-rounded"
      style="position: relative;"
    >
      <div
        class="Component-horizontalScrollContainer-7"
        style="overflow-x: auto; position: relative;"
      >
        <div>
          <div
            style="overflow-y: auto;"
          >
            <div>
              <table
                class="MuiTable-root"
                style="table-layout: auto;"
              >
                <thead
                  class="MuiTableHead-root"
                >
                  <tr
                    class="MuiTableRow-root MuiTableRow-head"
                  >
                    <th
                      class="MuiTableCell-root MuiTableCell-head MTableHeader-header-8 MuiTableCell-paddingNone"
                      scope="col"
                      style="font-weight: 700;"
                    />
                    <th
                      class="MuiTableCell-root MuiTableCell-head MTableHeader-header-8 MuiTableCell-alignLeft"
                      scope="col"
                      style="font-weight: 700; box-sizing: border-box;"
                    >
                      Metric
                    </th>
                    <th
                      class="MuiTableCell-root MuiTableCell-head MTableHeader-header-8 MuiTableCell-alignLeft"
                      scope="col"
                      style="font-weight: 700; box-sizing: border-box;"
                    >
                      Attribution window
                    </th>
                    <th
                      class="MuiTableCell-root MuiTableCell-head MTableHeader-header-8 MuiTableCell-alignLeft"
                      scope="col"
                      style="font-weight: 700; box-sizing: border-box;"
                    >
                      Recommendation
                    </th>
                  </tr>
                </thead>
                <tbody
                  class="MuiTableBody-root"
                >
                  <tr
                    class="MuiTableRow-root MuiTableRow-hover"
                    index="0"
                    level="0"
                    path="0"
                    style="transition: all ease 300ms; opacity: 0.8; cursor: pointer;"
                  >
                    <td
                      class="MuiTableCell-root MuiTableCell-body MuiTableCell-paddingNone"
                    >
                      <div
                        style="width: 42px; text-align: center; display: flex;"
                      >
                        <button
                          class="MuiButtonBase-root MuiIconButton-root Mui-disabled Mui-disabled"
                          disabled=""
                          style="transition: all ease 200ms; transform: none;"
                          tabindex="-1"
                          type="button"
                        >
                          <span
                            class="MuiIconButton-label"
                          >
                            <span
                              aria-hidden="true"
                              class="material-icons MuiIcon-root"
                            >
                              chevron_right
                            </span>
                          </span>
                        </button>
                      </div>
                    </td>
                    <td
                      class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignLeft"
                      style="box-sizing: border-box; font-family: 'Roboto Mono', monospace; font-weight: 600;"
                    >
                      metric_1
                       
                      <div
                        aria-disabled="true"
                        class="MuiChip-root makeStyles-primaryChip-2 MuiChip-outlined Mui-disabled"
                      >
                        <span
                          class="MuiChip-label"
                        >
                          Primary
                        </span>
                      </div>
                    </td>
                    <td
                      class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignLeft"
                      style="box-sizing: border-box; font-family: 'Roboto Mono', monospace;"
                    >
                      1 week
                    </td>
                    <td
                      class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignLeft"
                      style="box-sizing: border-box; font-family: 'Roboto Mono', monospace;"
                    >
                      Manual analysis required
                    </td>
                  </tr>
                  <tr
                    class="MuiTableRow-root MuiTableRow-hover"
                    index="1"
                    level="0"
                    path="1"
                    style="transition: all ease 300ms; opacity: 0.8; cursor: pointer;"
                  >
                    <td
                      class="MuiTableCell-root MuiTableCell-body MuiTableCell-paddingNone"
                    >
                      <div
                        style="width: 42px; text-align: center; display: flex;"
                      >
                        <button
                          class="MuiButtonBase-root MuiIconButton-root Mui-disabled Mui-disabled"
                          disabled=""
                          style="transition: all ease 200ms; transform: none;"
                          tabindex="-1"
                          type="button"
                        >
                          <span
                            class="MuiIconButton-label"
                          >
                            <span
                              aria-hidden="true"
                              class="material-icons MuiIcon-root"
                            >
                              chevron_right
                            </span>
                          </span>
                        </button>
                      </div>
                    </td>
                    <td
                      class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignLeft"
                      style="box-sizing: border-box; font-family: 'Roboto Mono', monospace; font-weight: 600;"
                    >
                      metric_2
                       
                    </td>
                    <td
                      class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignLeft"
                      style="box-sizing: border-box; font-family: 'Roboto Mono', monospace;"
                    >
                      1 hour
                    </td>
                    <td
                      class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignLeft"
                      style="box-sizing: border-box; font-family: 'Roboto Mono', monospace;"
                    >
                      Not analyzed yet
                    </td>
                  </tr>
                  <tr
                    class="MuiTableRow-root MuiTableRow-hover"
                    index="2"
                    level="0"
                    path="2"
                    style="transition: all ease 300ms; opacity: 0.8; cursor: pointer;"
                  >
                    <td
                      class="MuiTableCell-root MuiTableCell-body MuiTableCell-paddingNone"
                    >
                      <div
                        style="width: 42px; text-align: center; display: flex;"
                      >
                        <button
                          class="MuiButtonBase-root MuiIconButton-root Mui-disabled Mui-disabled"
                          disabled=""
                          style="transition: all ease 200ms; transform: none;"
                          tabindex="-1"
                          type="button"
                        >
                          <span
                            class="MuiIconButton-label"
                          >
                            <span
                              aria-hidden="true"
                              class="material-icons MuiIcon-root"
                            >
                              chevron_right
                            </span>
                          </span>
                        </button>
                      </div>
                    </td>
                    <td
                      class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignLeft"
                      style="box-sizing: border-box; font-family: 'Roboto Mono', monospace; font-weight: 600;"
                    >
                      metric_2
                       
                    </td>
                    <td
                      class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignLeft"
                      style="box-sizing: border-box; font-family: 'Roboto Mono', monospace;"
                    >
                      4 weeks
                    </td>
                    <td
                      class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignLeft"
                      style="box-sizing: border-box; font-family: 'Roboto Mono', monospace;"
                    >
                      Not analyzed yet
                    </td>
                  </tr>
                  <tr
                    class="MuiTableRow-root MuiTableRow-hover"
                    index="3"
                    level="0"
                    path="3"
                    style="transition: all ease 300ms; opacity: 0.8; cursor: pointer;"
                  >
                    <td
                      class="MuiTableCell-root MuiTableCell-body MuiTableCell-paddingNone"
                    >
                      <div
                        style="width: 42px; text-align: center; display: flex;"
                      >
                        <button
                          class="MuiButtonBase-root MuiIconButton-root"
                          style="transition: all ease 200ms; transform: none;"
                          tabindex="0"
                          type="button"
                        >
                          <span
                            class="MuiIconButton-label"
                          >
                            <span
                              aria-hidden="true"
                              class="material-icons MuiIcon-root"
                            >
                              chevron_right
                            </span>
                          </span>
                          <span
                            class="MuiTouchRipple-root"
                          />
                        </button>
                      </div>
                    </td>
                    <td
                      class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignLeft"
                      style="box-sizing: border-box; font-family: 'Roboto Mono', monospace; font-weight: 600;"
                    >
                      metric_3
                       
                    </td>
                    <td
                      class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignLeft"
                      style="box-sizing: border-box; font-family: 'Roboto Mono', monospace;"
                    >
                      6 hours
                    </td>
                    <td
                      class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignLeft"
                      style="box-sizing: border-box; font-family: 'Roboto Mono', monospace;"
                    >
                      Deploy 
                      test
                    </td>
                  </tr>
                </tbody>
              </table>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`renders the condensed table with some analyses in non-debug mode for a Conversion Metric 2`] = `
<div
  class="MuiTableContainer-root makeStyles-root-9 analysis-detail-panel"
>
  <table
    class="MuiTable-root"
  >
    <tbody
      class="MuiTableBody-root"
    >
      <tr
        class="MuiTableRow-root"
      >
        <th
          class="MuiTableCell-root MuiTableCell-head"
          role="cell"
          scope="row"
        >
          Last analyzed
        </th>
        <td
          class="MuiTableCell-root MuiTableCell-body"
        >
          <span
            class="makeStyles-root-14"
            title="09/05/2020, 20:00:00"
          >
            2020-05-10
          </span>
        </td>
      </tr>
      <tr
        class="MuiTableRow-root"
      >
        <th
          class="MuiTableCell-root MuiTableCell-head"
          role="cell"
          scope="row"
        >
          Analysis strategy
        </th>
        <td
          class="MuiTableCell-root MuiTableCell-body makeStyles-dataCell-11"
        >
          Exposed without crossovers and spammers
        </td>
      </tr>
      <tr
        class="MuiTableRow-root"
      >
        <th
          class="MuiTableCell-root MuiTableCell-head"
          role="cell"
          scope="row"
        >
          Analyzed participants
        </th>
        <td
          class="MuiTableCell-root MuiTableCell-body makeStyles-dataCell-11"
        >
          1400
           (
          140
           not final
          <span>
            ; 
            840
             in 
            control
          </span>
          <span>
            ; 
            560
             in 
            test
          </span>
          )
        </td>
      </tr>
      <tr
        class="MuiTableRow-root"
      >
        <th
          class="MuiTableCell-root MuiTableCell-head"
          role="cell"
          scope="row"
        >
          Difference interval
        </th>
        <td
          class="MuiTableCell-root MuiTableCell-body makeStyles-dataCell-11"
        >
          [
          -0.01
          ,
          0.01
          ]
        </td>
      </tr>
      <tr
        class="MuiTableRow-root"
      >
        <th
          class="MuiTableCell-root MuiTableCell-head"
          role="cell"
          scope="row"
        >
          Warnings
        </th>
        <td
          class="MuiTableCell-root MuiTableCell-body makeStyles-dataCell-11"
        >
          <div>
            Experiment period is too short. Wait a few days to be safer.
          </div>
          <div>
            The CI is too wide in comparison to the ROPE. Collect more data to be safer.
          </div>
        </td>
      </tr>
    </tbody>
  </table>
  <div
    class="makeStyles-plots-12"
  />
</div>
`;

exports[`renders the condensed table with some analyses in non-debug mode for a Conversion Metric 3`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "className": "makeStyles-plot-13",
        "data": Array [
          Object {
            "line": Object {
              "color": "#1f78b488",
            },
            "mode": "lines",
            "name": "test: lower bound",
            "type": "scatter",
            "x": Array [
              "2020-05-10T00:00:00.000Z",
            ],
            "y": Array [
              -112.3,
            ],
          },
          Object {
            "line": Object {
              "color": "#1f78b488",
            },
            "mode": "lines",
            "name": "test: upper bound",
            "type": "scatter",
            "x": Array [
              "2020-05-10T00:00:00.000Z",
            ],
            "y": Array [
              100,
            ],
          },
          Object {
            "line": Object {
              "color": "#ff7f0088",
            },
            "mode": "lines",
            "name": "control: lower bound",
            "type": "scatter",
            "x": Array [
              "2020-05-10T00:00:00.000Z",
            ],
            "y": Array [
              0,
            ],
          },
          Object {
            "line": Object {
              "color": "#ff7f0088",
            },
            "mode": "lines",
            "name": "control: upper bound",
            "type": "scatter",
            "x": Array [
              "2020-05-10T00:00:00.000Z",
            ],
            "y": Array [
              1000,
            ],
          },
        ],
        "debug": false,
        "layout": Object {
          "autosize": true,
          "hoverlabel": Object {
            "namelength": -1,
          },
          "margin": Object {
            "b": 48,
            "l": 32,
            "r": 16,
            "t": 64,
          },
          "showlegend": false,
          "title": "Conversion rate estimates by variation [%]",
        },
        "style": Object {
          "display": "inline-block",
          "position": "relative",
        },
        "useResizeHandler": false,
      },
      Object {},
    ],
    Array [
      Object {
        "className": "makeStyles-plot-13",
        "data": Array [
          Object {
            "line": Object {
              "width": 0,
            },
            "marker": Object {
              "color": "444",
            },
            "mode": "lines",
            "name": "difference: lower bound",
            "type": "scatter",
            "x": Array [
              "2020-05-10T00:00:00.000Z",
            ],
            "y": Array [
              -1,
            ],
          },
          Object {
            "fill": "tonexty",
            "fillcolor": "rgba(0,0,0,.2)",
            "line": Object {
              "width": 0,
            },
            "marker": Object {
              "color": "444",
            },
            "mode": "lines",
            "name": "difference: upper bound",
            "type": "scatter",
            "x": Array [
              "2020-05-10T00:00:00.000Z",
            ],
            "y": Array [
              1,
            ],
          },
          Object {
            "line": Object {
              "color": "rgba(0,0,0,.4)",
              "dash": "dash",
            },
            "mode": "lines",
            "name": "ROPE: lower bound",
            "type": "scatter",
            "x": Array [
              "2020-05-10T00:00:00.000Z",
            ],
            "y": Array [
              -1200,
            ],
          },
          Object {
            "line": Object {
              "color": "rgba(0,0,0,.4)",
              "dash": "dash",
            },
            "mode": "lines",
            "name": "ROPE: upper bound",
            "type": "scatter",
            "x": Array [
              "2020-05-10T00:00:00.000Z",
            ],
            "y": Array [
              1200,
            ],
          },
        ],
        "debug": false,
        "layout": Object {
          "autosize": true,
          "hoverlabel": Object {
            "namelength": -1,
          },
          "margin": Object {
            "b": 48,
            "l": 32,
            "r": 16,
            "t": 64,
          },
          "showlegend": false,
          "title": "Conversion rate difference estimates [%]",
        },
        "style": Object {
          "display": "inline-block",
          "position": "relative",
        },
        "useResizeHandler": false,
      },
      Object {},
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`renders the condensed table with some analyses in non-debug mode for a Revenue Metric 1`] = `
<div
  class="analysis-latest-results"
>
  <div
    class="makeStyles-root-15"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation2 MuiPaper-rounded"
      style="position: relative;"
    >
      <div
        class="Component-horizontalScrollContainer-21"
        style="overflow-x: auto; position: relative;"
      >
        <div>
          <div
            style="overflow-y: auto;"
          >
            <div>
              <table
                class="MuiTable-root"
                style="table-layout: auto;"
              >
                <thead
                  class="MuiTableHead-root"
                >
                  <tr
                    class="MuiTableRow-root MuiTableRow-head"
                  >
                    <th
                      class="MuiTableCell-root MuiTableCell-head MTableHeader-header-22 MuiTableCell-paddingNone"
                      scope="col"
                      style="font-weight: 700;"
                    />
                    <th
                      class="MuiTableCell-root MuiTableCell-head MTableHeader-header-22 MuiTableCell-alignLeft"
                      scope="col"
                      style="font-weight: 700; box-sizing: border-box;"
                    >
                      Metric
                    </th>
                    <th
                      class="MuiTableCell-root MuiTableCell-head MTableHeader-header-22 MuiTableCell-alignLeft"
                      scope="col"
                      style="font-weight: 700; box-sizing: border-box;"
                    >
                      Attribution window
                    </th>
                    <th
                      class="MuiTableCell-root MuiTableCell-head MTableHeader-header-22 MuiTableCell-alignLeft"
                      scope="col"
                      style="font-weight: 700; box-sizing: border-box;"
                    >
                      Recommendation
                    </th>
                  </tr>
                </thead>
                <tbody
                  class="MuiTableBody-root"
                >
                  <tr
                    class="MuiTableRow-root MuiTableRow-hover"
                    index="0"
                    level="0"
                    path="0"
                    style="transition: all ease 300ms; opacity: 0.8; cursor: pointer;"
                  >
                    <td
                      class="MuiTableCell-root MuiTableCell-body MuiTableCell-paddingNone"
                    >
                      <div
                        style="width: 42px; text-align: center; display: flex;"
                      >
                        <button
                          class="MuiButtonBase-root MuiIconButton-root Mui-disabled Mui-disabled"
                          disabled=""
                          style="transition: all ease 200ms; transform: none;"
                          tabindex="-1"
                          type="button"
                        >
                          <span
                            class="MuiIconButton-label"
                          >
                            <span
                              aria-hidden="true"
                              class="material-icons MuiIcon-root"
                            >
                              chevron_right
                            </span>
                          </span>
                        </button>
                      </div>
                    </td>
                    <td
                      class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignLeft"
                      style="box-sizing: border-box; font-family: 'Roboto Mono', monospace; font-weight: 600;"
                    >
                      metric_1
                       
                      <div
                        aria-disabled="true"
                        class="MuiChip-root makeStyles-primaryChip-16 MuiChip-outlined Mui-disabled"
                      >
                        <span
                          class="MuiChip-label"
                        >
                          Primary
                        </span>
                      </div>
                    </td>
                    <td
                      class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignLeft"
                      style="box-sizing: border-box; font-family: 'Roboto Mono', monospace;"
                    >
                      1 week
                    </td>
                    <td
                      class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignLeft"
                      style="box-sizing: border-box; font-family: 'Roboto Mono', monospace;"
                    >
                      Manual analysis required
                    </td>
                  </tr>
                  <tr
                    class="MuiTableRow-root MuiTableRow-hover"
                    index="1"
                    level="0"
                    path="1"
                    style="transition: all ease 300ms; opacity: 0.8; cursor: pointer;"
                  >
                    <td
                      class="MuiTableCell-root MuiTableCell-body MuiTableCell-paddingNone"
                    >
                      <div
                        style="width: 42px; text-align: center; display: flex;"
                      >
                        <button
                          class="MuiButtonBase-root MuiIconButton-root Mui-disabled Mui-disabled"
                          disabled=""
                          style="transition: all ease 200ms; transform: none;"
                          tabindex="-1"
                          type="button"
                        >
                          <span
                            class="MuiIconButton-label"
                          >
                            <span
                              aria-hidden="true"
                              class="material-icons MuiIcon-root"
                            >
                              chevron_right
                            </span>
                          </span>
                        </button>
                      </div>
                    </td>
                    <td
                      class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignLeft"
                      style="box-sizing: border-box; font-family: 'Roboto Mono', monospace; font-weight: 600;"
                    >
                      metric_2
                       
                    </td>
                    <td
                      class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignLeft"
                      style="box-sizing: border-box; font-family: 'Roboto Mono', monospace;"
                    >
                      1 hour
                    </td>
                    <td
                      class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignLeft"
                      style="box-sizing: border-box; font-family: 'Roboto Mono', monospace;"
                    >
                      Not analyzed yet
                    </td>
                  </tr>
                  <tr
                    class="MuiTableRow-root MuiTableRow-hover"
                    index="2"
                    level="0"
                    path="2"
                    style="transition: all ease 300ms; opacity: 0.8; cursor: pointer;"
                  >
                    <td
                      class="MuiTableCell-root MuiTableCell-body MuiTableCell-paddingNone"
                    >
                      <div
                        style="width: 42px; text-align: center; display: flex;"
                      >
                        <button
                          class="MuiButtonBase-root MuiIconButton-root Mui-disabled Mui-disabled"
                          disabled=""
                          style="transition: all ease 200ms; transform: none;"
                          tabindex="-1"
                          type="button"
                        >
                          <span
                            class="MuiIconButton-label"
                          >
                            <span
                              aria-hidden="true"
                              class="material-icons MuiIcon-root"
                            >
                              chevron_right
                            </span>
                          </span>
                        </button>
                      </div>
                    </td>
                    <td
                      class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignLeft"
                      style="box-sizing: border-box; font-family: 'Roboto Mono', monospace; font-weight: 600;"
                    >
                      metric_2
                       
                    </td>
                    <td
                      class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignLeft"
                      style="box-sizing: border-box; font-family: 'Roboto Mono', monospace;"
                    >
                      4 weeks
                    </td>
                    <td
                      class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignLeft"
                      style="box-sizing: border-box; font-family: 'Roboto Mono', monospace;"
                    >
                      Not analyzed yet
                    </td>
                  </tr>
                  <tr
                    class="MuiTableRow-root MuiTableRow-hover"
                    index="3"
                    level="0"
                    path="3"
                    style="transition: all ease 300ms; opacity: 0.8; cursor: pointer;"
                  >
                    <td
                      class="MuiTableCell-root MuiTableCell-body MuiTableCell-paddingNone"
                    >
                      <div
                        style="width: 42px; text-align: center; display: flex;"
                      >
                        <button
                          class="MuiButtonBase-root MuiIconButton-root"
                          style="transition: all ease 200ms; transform: none;"
                          tabindex="0"
                          type="button"
                        >
                          <span
                            class="MuiIconButton-label"
                          >
                            <span
                              aria-hidden="true"
                              class="material-icons MuiIcon-root"
                            >
                              chevron_right
                            </span>
                          </span>
                          <span
                            class="MuiTouchRipple-root"
                          />
                        </button>
                      </div>
                    </td>
                    <td
                      class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignLeft"
                      style="box-sizing: border-box; font-family: 'Roboto Mono', monospace; font-weight: 600;"
                    >
                      metric_3
                       
                    </td>
                    <td
                      class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignLeft"
                      style="box-sizing: border-box; font-family: 'Roboto Mono', monospace;"
                    >
                      6 hours
                    </td>
                    <td
                      class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignLeft"
                      style="box-sizing: border-box; font-family: 'Roboto Mono', monospace;"
                    >
                      Deploy 
                      test
                    </td>
                  </tr>
                </tbody>
              </table>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`renders the condensed table with some analyses in non-debug mode for a Revenue Metric 2`] = `
<div
  class="MuiTableContainer-root makeStyles-root-23 analysis-detail-panel"
>
  <table
    class="MuiTable-root"
  >
    <tbody
      class="MuiTableBody-root"
    >
      <tr
        class="MuiTableRow-root"
      >
        <th
          class="MuiTableCell-root MuiTableCell-head"
          role="cell"
          scope="row"
        >
          Last analyzed
        </th>
        <td
          class="MuiTableCell-root MuiTableCell-body"
        >
          <span
            class="makeStyles-root-28"
            title="09/05/2020, 20:00:00"
          >
            2020-05-10
          </span>
        </td>
      </tr>
      <tr
        class="MuiTableRow-root"
      >
        <th
          class="MuiTableCell-root MuiTableCell-head"
          role="cell"
          scope="row"
        >
          Analysis strategy
        </th>
        <td
          class="MuiTableCell-root MuiTableCell-body makeStyles-dataCell-25"
        >
          Exposed without crossovers and spammers
        </td>
      </tr>
      <tr
        class="MuiTableRow-root"
      >
        <th
          class="MuiTableCell-root MuiTableCell-head"
          role="cell"
          scope="row"
        >
          Analyzed participants
        </th>
        <td
          class="MuiTableCell-root MuiTableCell-body makeStyles-dataCell-25"
        >
          1400
           (
          140
           not final
          <span>
            ; 
            840
             in 
            control
          </span>
          <span>
            ; 
            560
             in 
            test
          </span>
          )
        </td>
      </tr>
      <tr
        class="MuiTableRow-root"
      >
        <th
          class="MuiTableCell-root MuiTableCell-head"
          role="cell"
          scope="row"
        >
          Difference interval
        </th>
        <td
          class="MuiTableCell-root MuiTableCell-body makeStyles-dataCell-25"
        >
          [
          -0.01
          ,
          0.01
          ]
        </td>
      </tr>
      <tr
        class="MuiTableRow-root"
      >
        <th
          class="MuiTableCell-root MuiTableCell-head"
          role="cell"
          scope="row"
        >
          Warnings
        </th>
        <td
          class="MuiTableCell-root MuiTableCell-body makeStyles-dataCell-25"
        >
          <div>
            Experiment period is too short. Wait a few days to be safer.
          </div>
          <div>
            The CI is too wide in comparison to the ROPE. Collect more data to be safer.
          </div>
        </td>
      </tr>
    </tbody>
  </table>
  <div
    class="makeStyles-plots-26"
  />
</div>
`;

exports[`renders the condensed table with some analyses in non-debug mode for a Revenue Metric 3`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "className": "makeStyles-plot-27",
        "data": Array [
          Object {
            "line": Object {
              "color": "#1f78b488",
            },
            "mode": "lines",
            "name": "test: lower bound",
            "type": "scatter",
            "x": Array [
              "2020-05-10T00:00:00.000Z",
            ],
            "y": Array [
              -1.123,
            ],
          },
          Object {
            "line": Object {
              "color": "#1f78b488",
            },
            "mode": "lines",
            "name": "test: upper bound",
            "type": "scatter",
            "x": Array [
              "2020-05-10T00:00:00.000Z",
            ],
            "y": Array [
              1,
            ],
          },
          Object {
            "line": Object {
              "color": "#ff7f0088",
            },
            "mode": "lines",
            "name": "control: lower bound",
            "type": "scatter",
            "x": Array [
              "2020-05-10T00:00:00.000Z",
            ],
            "y": Array [
              0,
            ],
          },
          Object {
            "line": Object {
              "color": "#ff7f0088",
            },
            "mode": "lines",
            "name": "control: upper bound",
            "type": "scatter",
            "x": Array [
              "2020-05-10T00:00:00.000Z",
            ],
            "y": Array [
              10,
            ],
          },
        ],
        "debug": false,
        "layout": Object {
          "autosize": true,
          "hoverlabel": Object {
            "namelength": -1,
          },
          "margin": Object {
            "b": 48,
            "l": 32,
            "r": 16,
            "t": 64,
          },
          "showlegend": false,
          "title": "Revenue estimates by variation [$]",
        },
        "style": Object {
          "display": "inline-block",
          "position": "relative",
        },
        "useResizeHandler": false,
      },
      Object {},
    ],
    Array [
      Object {
        "className": "makeStyles-plot-27",
        "data": Array [
          Object {
            "line": Object {
              "width": 0,
            },
            "marker": Object {
              "color": "444",
            },
            "mode": "lines",
            "name": "difference: lower bound",
            "type": "scatter",
            "x": Array [
              "2020-05-10T00:00:00.000Z",
            ],
            "y": Array [
              -0.01,
            ],
          },
          Object {
            "fill": "tonexty",
            "fillcolor": "rgba(0,0,0,.2)",
            "line": Object {
              "width": 0,
            },
            "marker": Object {
              "color": "444",
            },
            "mode": "lines",
            "name": "difference: upper bound",
            "type": "scatter",
            "x": Array [
              "2020-05-10T00:00:00.000Z",
            ],
            "y": Array [
              0.01,
            ],
          },
          Object {
            "line": Object {
              "color": "rgba(0,0,0,.4)",
              "dash": "dash",
            },
            "mode": "lines",
            "name": "ROPE: lower bound",
            "type": "scatter",
            "x": Array [
              "2020-05-10T00:00:00.000Z",
            ],
            "y": Array [
              -12,
            ],
          },
          Object {
            "line": Object {
              "color": "rgba(0,0,0,.4)",
              "dash": "dash",
            },
            "mode": "lines",
            "name": "ROPE: upper bound",
            "type": "scatter",
            "x": Array [
              "2020-05-10T00:00:00.000Z",
            ],
            "y": Array [
              12,
            ],
          },
        ],
        "debug": false,
        "layout": Object {
          "autosize": true,
          "hoverlabel": Object {
            "namelength": -1,
          },
          "margin": Object {
            "b": 48,
            "l": 32,
            "r": 16,
            "t": 64,
          },
          "showlegend": false,
          "title": "Revenue difference estimates [$]",
        },
        "style": Object {
          "display": "inline-block",
          "position": "relative",
        },
        "useResizeHandler": false,
      },
      Object {},
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
